sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
 
     rng = np.random.RandomState(0)
     X = rng.rand(20, 10)
-    for n_clus in [-1, 0]:
-        agc = AgglomerativeClustering(n_clusters=n_clus)
-        msg = ("n_clusters should be an integer greater than 0."
-               " %s was provided." % str(agc.n_clusters))
-        assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = -1
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("n_clusters should be an integer greater than 0."
+           " %s was provided." % str(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = 0
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
```

Comment:
Because of the order of the checks, the first check is general and checks for None and 0. I did this because None computes to less than 0 in Python 2 and not in Python 3. 

I can swap them around and add a check for None when checking if `n_clusters <=0` and this should give the same message for both cases.